### PR TITLE
Enhance fakepg: alltoall and alltoall_base

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107626
* #107625
* __->__ #107624

from working on a starter task with @wanchaol (T161350434):
Enhance Fake Process Group by adding missing collectives: alltoall, alltoall_base

Differential Revision: [D48606654](https://our.internmc.facebook.com/intern/diff/D48606654)

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
lgtm

File: torch/testing/_internal/distributed/fake_pg.py [hunk 1]
```diff
@@ -86,6 +87,24 @@ def scatter(
...
 
+    def alltoall(
+        self,
+        output_tensors: List[Tensor],
+        input_tensors: List[Tensor],
+        opts=AllToAllOptions(),
+    ):
+        return ret_work(output_tensors)
+
+    def alltoall_base(
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/testing/_internal/distributed/fake_pg.py:98]: I think the `alltoall_base` is from the ProcessGroup C++ API, and this is the API custom PG need to override. python `alltoall_single` calls into `alltoall_base`

[the remark this answers is not in the document]
Comment (xmfan) [reply torch/testing/_internal/distributed/fake_pg.py:98]: That's right, all_to_all_single's implementation calls into the ProcessGroup's alltoall_base method: https://github.com/pytorch/pytorch/blob/cbcd55104504609e7b42d208310514a9e8afb2c0/torch/distributed/distributed_c10d.py#L3523-L3527

**Comment by xmfan** [comment]
@xmfan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48606654).

**Pull request merged**